### ISSUE TITLE
Tera doesn't honor auto-escape settings of included templates

### ISSUE THREAD

**Issue by d4h0** [Keats/tera#595]
Hi,

I have the following code in a `.tera` file:

```
        {% filter markdown %}
            {% include "__markdown__.md" %}
        {% endfilter %}
```

Auto-escape is activated for `.tera` files, but not for `.md` files:

`tera.autoescape_on(vec!["tera"])`

But values in the Markdown file are auto-escaped:

If the `layout` variable is "layouts/article.tera", `# test: {{layout}` in a Markdown file leads to `# test: layouts&#x2F;article.tera` being passed to my Markdown filter

Disabling auto-escape for `.tera` files, or using the safe filter in the Markdown file prevents the escape.

Is this a bug? I was expecting that templates with a `.md` extension would not be auto-escaped.

**Comment by d4h0** [comment]
Normally, I could use `Tera::render_str` to render the Markdown and return the result from my Markdown filter. But unfortunately [Tera::render_str is missing the auto-escape option](https://github.com/Keats/tera/issues/559) at the moment.

Is there another workaround?

For now, I use `Tera::one_off`, however now my Markdown doesn't have access to the other templates.

**Comment by Keats** [comment]
That's a bug! It shouldn't escape the content.

**Comment by Keats** [comment]
Oh I see what's happening. The included file is correctly not escaped but the filter i, because it's in an HTML file. I'm not sure if it works but you can try:
```
{% filter safe %}
        {% filter markdown %}
            {% include "__markdown__.md" %}
        {% endfilter %}
{% endfilter %}
```
? If it doesn't then I don't think there is a way to do it currently.

**Comment by d4h0** [comment]
Thanks for having a looks again!

To test this, I changed my code to this:

```
                {% filter safe %}
                    {# filter markdown #}
                    {% include "__markdown__.md" %}
                    {# endfilter #}
                {% endfilter %
```

(As mentioned, my Markdown parser removes unnesesery HTML escapes, so I deactivated my Markdown filter).

However, I get this error:

```
Error while generating page:
Failed to render 'layouts/article.tera'

Caused by:
    Filter 'safe' not found

Location:
    ssg/src/lib.rs:881:49
```

Is the `safe` filter available as block filter?

**Comment by Keats** [comment]
The safe filter doesn't actually exist. With the commit above it should work I think? If you can try it that would be great

**Comment by d4h0** [comment]
So, I installed `tera` from the commit above via:

`tera = { git = "https://github.com/Keats/tera.git", branch = "next", ref = "1fa659da5fcf118d3fea309a71726583171f2777" }`

That's correct, right?

This seems to work (the slash isn't escaped anymore).

However, this messed up my templates pretty badly. It seems the `VSCode` extension has problems with this commit. I didn't yet look at my template fully, but it seems some stuff was just deleted.

I'm using a custom filter like this:

`<div class="{{ "mt-24 sm:mt-16 px-4 max-w-screen-lg mx-auto flex flex-col items-center text-center"|tw}}">`

Which seems to lead to problems, it seems. Some classes were just deleted.

I'll switch to `class='{{ "xy"|tw}}'`, but someone probably should have a look at this problem.

Or maybe it wasn't this commit, but a new version of the `VSCode` extension. I didn't work on my templates for some time.

The problem this issue is about seems to be fixed, but my templates are so messed up, that I'm not 100% sure. If there is still a problem after I fixed my templates (of course I didn't commit them to git... 😆), I will report back here.

**Comment by Keats** [comment]
Hm none of the changes in that branch should cause issues like that.

**Comment by d4h0** [comment]
Very strange. The VSCode extensions also wasn't [updated since last year](https://marketplace.visualstudio.com/items?itemName=karunamurti.tera). I just saw, that the extension wasn't created by you, btw.

Maybe it was some other extension that handles HTML. Anyway, hopefully that will not happen again, after I switch to the safer syntax I mentioned above.

---

I just realized something:

It doesn't matter if I use:

```
            {% filter safe %}
            {% include "__markdown__.md" %}
            {% endfilter %
```

or just

```
            {% include "__markdown__.md" %}
```

...variables are, correctly, not escaped!

But if I do this:

```
            {% filter markdown %}
            {% include "__markdown__.md" %}
            {% endfilter %
```

...and print out the string, which `markdown` receives from `tera`, then variables are escaped (`layouts/article.tera` becomes `layouts&#x2F;article.tera`).

**Comment by Keats** [comment]
Yep, this should be fixed in the branch in theory.
The first 2 examples work as expected: the include is not auto-escaped because the extensions doesn't match what should the extensions. However the last one is running the markdown filter which outputs in the .tera file which is auto-escaped and thus require `safe`.
This is a bit weird but I believe Jinja2 works the same way.

**Comment by d4h0** [comment]
I'm not sure, if I understand this correctly.

So my Markdown filter implements `is_safe`, which returns `true`. So nothing my filter outputs should be escaped.

If an included template doesn't escape variables normally, this should be true as well, if this `include` is within a filter block, I believe (the filter should get, what otherwise would be included).

I'd expect, that variables within `__markdown__.md` are expanded before the content is included (my first two examples indicate, that this is true. Variables are not escaped here), so there wouldn't be anything in the parent template (from where I call `include` and `filter`) to escape.

So basically, wrapping `{% include "__markdown__.md" %}` with my Markdown filter changes that.

Do I miss anything?

**Comment by Keats** [comment]
> So my Markdown filter implements is_safe, which returns true. 

Ah yes in that case nothing should be escaped, unless there is a bug with `is_safe` for filter sections.
